### PR TITLE
Add contributing documentation

### PR DESCRIPTION
**Pull request by max-Hawkins**
Given the recent interest in contributions to Metal.jl, I thought it'd be good to put some more explicit information in the docs. I wasn't sure what all to include. Should I include some sort of explanation for the CMT naming scheme ('Mtl' vs 'MTL' vs 'Mt'...etc)? Maybe I should give some pointers on Objective-C?

### CODE REVIEW

File: docs/src/contributing.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+When new functionality is added to the low-level CMT library, it will
+need to be rebuilt. This is done by running the `deps/build_cmt.jl` script. If this
+successfully builds, you can move onto the next step: rewrapping CMT to include the added
+functionality. From the package's top-level directory, the command is
+`julia --project=deps deps/build_cmt.jl`
+
+### Wrapping
+
+With a succesfully-built CMT library, you'll need to re-wrap it to include the new
+functionality. From the deps folder, run `julia --project=. wrap.jl`
```
Comment (maleadt) [inline docs/src/contributing.md:103]: it's in the `res` folder

**Comment by maleadt** [comment]
I think this is a pretty good start, and wouldn't bog it down with more ObjC details. Most users are going to be focusing on the Julia level anyway, so instead it may be better to link to or add a bit on kernel programming too?

**Comment by habemus-papadum** [comment]
@max-Hawkins this is really great.  One thing that might be useful to beginners is a quick code snippet on how to run only a subset of the tests (I don't actually know the answer and will be digging around in just a second).  In general the testing infrastructure seems really quite impressive and any insight into the basic mechanics and capabilities would be interesting and helpful.

**Comment by maleadt** [comment]
Sorry, forgot about this.

> What aspects of kernel programming? Just curious what you had in mind.

Nothing much, just a simple `vadd` kernel to show the core principles?

---

Couple of tidbits about the test system: The system actually partially behaves like a script, so it's often easiest to invoke it directly:

```
❯ julia --project=test test/runtests.jl --help
Usage: runtests.jl [--help] [--list] [--jobs=N] [TESTS...]

       --help             Show this text.
       --list             List all available tests.
       --quickfail        Fail the entire run as soon as a single test errored.
       --jobs=N           Launch `N` processes to perform tests (default: Threads.nthreads()).

       Remaining arguments filter the tests that will be executed.
```

Since you might need additional test dependencies, this can also be accomplished through Pkg:

```
❯ julia --project -e 'using Pkg; Pkg.test(; test_args=`--help`)'
     Testing Metal
      Status `/private/var/folders/q0/wbdz2dpd0q9gtjk3wkj1fcnh0000gn/T/jl_Ci5y77/Project.toml`
...
     Testing Running tests...
Usage: runtests.jl [--help] [--list] [--jobs=N] [TESTS...]
```

Personally I have an alias that simplifies this -- but that probably doesn't belong in the docs:

```
function jltest {
    julia=(julia)

    # certain arguments (like those beginnning with a +) need to come first
    if [[ $# -gt 0 && "$1" = +* ]]; then
        julia+=("$1")
        shift
    fi

    "${julia[@]}" --startup-file=no --project -e "using Pkg; Pkg.API.test(; test_args=ARGS)" "$@"
}
```

Anyway, to run individual tests you can either pass it as an arg to the test system:

```
❯ julia --project=test test/runtests.jl metal
┌ Info: System information:
│ macOS 13.2.1, Darwin 21.5.0
│
│ Toolchain:
│ - Julia: 1.8.5
│ - LLVM: 13.0.1
│
│ 1 device:
└ - Apple M1 Pro (64.000 KiB allocated)
[ Info: Using cmt library from /Users/tim/Julia/depot/scratchspaces/dde4c033-4e86-420c-a63e-0dd931031962/cmt/lib
┌ Info: Using Metal LLVM back-end from /Users/tim/Julia/depot/artifacts/0ec307395e9dd683100b6ecbb14963c6ddd232e6/bin:
│ LLVM (http://llvm.org/):
│   LLVM version 13.0.1
│   Optimized build.
│   Default target: aarch64-apple-darwin22.3.0
└   Host CPU: cyclone
               |          | ---------------- CPU ---------------- |
Test  (Worker) | Time (s) | GC (s) | GC % | Alloc (MB) | RSS (MB) |
metal      (2) |        started at 2023-02-24T13:24:13.799
```

... or run the files separately after including `test/setup.jl` (again, for your information, probably doesn't belong in the docs):

```
❯ julia --project=test -L test/setup.jl test/metal.jl
```

The test system is used in CUDA.jl, oneAPI.jl and Metal.jl. It would probably be useful to extract it in a helper package, but I haven't found a good way to abstract the subtle differences between back-ends.

File: docs/src/contributing.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+for your new functionality. To do that, you can either pass the name of the testset to the
+`test/runtests.jl` script: `julia --project=test test/runtests.jl metal` or you can isolate test
+files by running them alone after running the `test/setup.jl` script:
+`julia --project=test -L test/setup.jl test/metal.jl`
+
+## Thank You and Good Luck
+
+Open-source projects like this only happen because people like you are willing to spend
+their free time helping out. Most anything you're able to do is helpful, but if you get
+stuck, seek guidance from Slack or Discord. Don't feel like your contribution has to be
```
Comment (maxwindiff) [inline docs/src/contributing.md:148]: Wait we have a discord?

File: docs/src/metal-programming.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+The most fundamental idea of programming GPUs (when compared to serial CPU programming) is
+its *parallelism*. A GPU function (kernel), when called, is not just ran once in isolation.
+Rather, numerous (often thousands to millions) psuedo-independent instances (called threads)
+of the kernel are executed in parallel. These threads are arranged in a hierarchy that allows
+for varying levels of synchronization. For Metal, the hierarchy is as follows:
+
+- Thread: A single execution unit of the kernel
+- Threadgroup: A collection of threads that share a common block of memory and synchronization
+barriers
+- Grid: A collection of threadgroups
```
Comment (maxwindiff) [inline docs/src/metal-programming.md:28]: Should we either change the explanation to match Apple's convention, or use a different terminology here?

File: docs/src/contributing.md (hunk 2 above)
Comment (maleadt) [reply docs/src/contributing.md:148]: Was probably intended to be Discourse :-)

File: docs/src/metal-programming.md (hunk 3 above)
Comment (maleadt) [reply docs/src/metal-programming.md:28]: Yeah let's wait for https://github.com/JuliaGPU/Metal.jl/pull/119 and update the docs then

File: docs/src/metal-programming.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+The threadgroup and grid sizes are set by the user when launching the GPU kernel. There are
+upper limits determined by the targeted hardware, and the sizes can be 1, 2, or 3-dimensional. For
+Metal.jl, these sizes are set using the `@metal` macro's keyword arguments. The `grid`
+keyword determines the grid size while the `threads` keyword determines the threadgroup size.
+
+For example, given a 10x10x3 image that you want to run a function independently on each pixel,
+the kernel launch code might look like the following:
+`@metal grid=(3) threads=(10,10) my_kernel(gpu_image_array)`
```
Comment (maleadt) [inline docs/src/metal-programming.md:37]: Needs to be adapted to the new syntax. Also could just specify `groups=3`, dropping the parenthesis.

File: docs/src/metal-programming.md (hunk 3 above)
Comment (maleadt) [reply docs/src/metal-programming.md:28]: Actually, let's keep `grid` here but mention `groups` below in the `@metal` invocation.

File: docs/src/metal-programming.md (hunk 4 above)
Comment (maleadt) [inline docs/src/metal-programming.md:37]:
```suggestion
`@metal threads=(10,10) groups=3 my_kernel(gpu_image_array)`
```

File: docs/src/metal-programming.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+Each thread has its own set of private variables. Most importantly, each thread has
+associated unique indices to identify itself within its threadgroup and grid.
+These are traditionally what are used to differentiate execution across threads. You can
+also query what the grid and threadgroup sizes are as well.
+
+For Metal.jl, these values are accessed via the following functions:
+- thread_index_in_threadgroup()
+
+- grid_size_Xd()
```
Comment (maleadt) [inline docs/src/metal-programming.md:57]:
```suggestion
For Metal.jl, these values are accessed via the following functions:

- thread_index_in_threadgroup()
- grid_size_Xd()
```

File: docs/src/contributing.md (hunk 2 above)
Comment (maleadt) [inline docs/src/contributing.md:148]:
```suggestion
stuck, seek guidance from Slack or Discourse. Don't feel like your contribution has to be
```

File: docs/src/contributing.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+successfully builds, you can move onto the next step: rewrapping CMT to include the added
+functionality. From the package's top-level directory, the command is
+`julia --project=deps deps/build_cmt.jl`
+
+### Wrapping
+
+With a succesfully-built CMT library, you'll need to re-wrap it to include the new
+functionality. From the `res` folder, run `julia --project=. wrap.jl`
+
+
```
Comment (maleadt) [inline docs/src/contributing.md:105]:
```suggestion
```

File: docs/src/contributing.md [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+If you're adding some functionality that originates from Metal Shading Language (MSL)
+directly (rather than high-level Julia functionality), the workflow will likely look like
+the below. If you're adding something that only relies on pure Julia additions, you will
+skip the first two steps.
+
+1. Create Objective-C module
+2. Create low-level, Julia wrappers for the Obj-C interface
+3. Create high-level Julia structures and functionality
+4. Create tests for added functionality
```
Comment (maleadt) [inline docs/src/contributing.md:40]:
```suggestion
1. Create low-level, Julia wrappers for the Obj-C interface
2. Create high-level Julia structures and functionality
3. Create tests for added functionality
```

**Pull request merged**